walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -22,59 +22,87 @@
-      return times; 
+      return timesForEveryone; 
     }
-    if (attendees.size() == 0) {
+    if (request.getAttendees().size() == 0 && request.getOptionalAttendees().size() == 0) {
       // If there are no attendees, the entire day is available to book a meeting.
-      times.add(TimeRange.WHOLE_DAY);
-      return times;
+      timesForEveryone.add(TimeRange.WHOLE_DAY);
+      return timesForEveryone;
     }
 
-    int currentEndTime = TimeRange.START_OF_DAY;
+    currentEndTime = TimeRange.START_OF_DAY;
     for (Event event : events) {
       // Ignores this event's time range if the people attending this event are 
       // not attending the requested meeting. 
       Set<String> eventAttendees = new HashSet<String>(event.getAttendees());
-      eventAttendees.retainAll(attendees);
-      if (eventAttendees.isEmpty()) {
-        continue;
-      }
-
-      TimeRange time = event.getWhen();
-      // Skips this event because it ended before or at the same time as
-      // the latest event.
-      if (time.end() <= currentEndTime) {
+      eventAttendees.retainAll(request.getAttendees());
+      Set<String> optionalEventAttendeesForRequestedMeeting = 
+          new HashSet<String>(event.getAttendees());
+      optionalEventAttendeesForRequestedMeeting.retainAll(request.getOptionalAttendees());
+      if (eventAttendees.isEmpty() && optionalEventAttendeesForRequestedMeeting.isEmpty()) {
         continue;
       }
-      if ((time.start() <= currentEndTime) || 
-          (request.getDuration() > (time.start() - currentEndTime)) {
-        // Handles events starting before but ending after the currentEndTime and the case 
-        // where the request meeeting duration is longer than the time gap.
-        currentEndTime = time.end();
-        continue;
-      }
-      TimeRange gapTimeRange = TimeRange.fromStartEnd(currentEndTime, time.start(), false);
-      currentEndTime = time.end();
-      times.add(gapTimeRange);
+      addTimeGapIfPossible(request, eventAttendees, event);
     }
-
+    
     // Adds a timerange for after the latest event ends.
     if (currentEndTime < TimeRange.END_OF_DAY) {
       TimeRange latestTimeRange = TimeRange.fromStartEnd(currentEndTime, 
           TimeRange.END_OF_DAY, true);
-      times.add(latestTimeRange);
+      timesForEveryone.add(latestTimeRange);
+    }
+
+    if (timesForEveryone.size() == 0 && request.getAttendees().size() != 0) {
+      return timesForConfirmed;
+    }
+    return timesForEveryone;
+  }
+
+  /**
+   * Adds a TimeRange to indicate a time gap for either {@code timesForEveryone} or {@code timesForConfirmed} depending
+   * on whether optional or confirmed attendees are attending the current {@code event}.
+   * 
+   * <p>Checks {@code request.duration} and {@code eventAttendees} to see if a time gap is possible for which people.
+   */
+  public void addTimeGapIfPossible(MeetingRequest request, Set<String> eventAttendees, Event event) {
```

Comment:
usually if you are setting a value for a class member you prepend the method name with 'set'. So for this I would rename 'setTimeGapIfPossible'.

Reply:
done